Website: Fleet Sandbox

eashaw opened the pull request:
Issue
- https://github.com/fleetdm/fleet/issues/5036


Changes (updated 07-26)
---

Changes to existing files:

- `User.js`
	- Added three new attributes to the User model:
		- `fleetSandboxURL`: The URL of the Fleet sandbox instance that was provisioned for this user (string)
		- `fleetSandboxExpiresAt`: A JS Timestamp of when the User's sandbox instance will expire. (number)
		- `fleetSandboxDemoKey`: The UUID that is used as the password of the User's Fleet Sandbox instance
	- Changed the `organization` attribute to be optional
- `signup.js`
	- Updated inputs:
		- Added one new input: `signupReason`: can either be "Try Fleet Sandbox" or "Buy a license"
			- defaults to "Buy a license"
		- Changed the `organization` input to be optional
	- Updated behavior:
		- Throws an error if sails.config.custom.cloudProvisionerSecret is not set.
		- Sends a POST request to the cloud provisioner API at `sandbox.fleetdm.com/new` to provision a new Fleet Sandbox instance for all users signing up
			- The POST request body includes:
				- `name`: The user's name (Note: for user's who sign up on the sandbox registration page, the user's first and last name will be parts of their email address)
				- `email`: The email address the user signed up with
				- `password`: A generated UUID that is stored on the User model as `fleetSandboxDemoKey`. This value cannot be changed.
				- `sandbox_expiration`: An ISO timestamp of when this user's Fleet Sandbox instance will expire. This is generated when the user signs up, and is set to 24 hours from the time it was generated. Note: We store this value as an attribute on the User model as a JS Timestamp called `fleetSandboxExpiresAt`
			- The response from the cloud provisioner will have a `URL` that has the url of the provisioned Fleet Sandbox Instance.
			- If the request fails, or the response data does not include a `URL` property, this action will return an error.
		- If `signupReason` is "Try Fleet Sandbox", this action will
			- Continuously poll the `/healthz` endpoint of the newly provisioned Fleet Sandbox instance until it returns a 200 response.
			- If 10 seconds pass and the Sandbox instance does not return a 200 response, an error is returned
		- If either of the requests fail, this action will return an error and will not save the new user in the database.
		- This action now saves `fleetSandboxURL`, `fleetSandboxDemoKey`, and `fleetSandboxExpiresAt` values to all new users created.
- `view-login`
	- If a logged in user has a billing card saved, they are redirected to the customer dashboard, otherwise, they are redirected to their Fleet Sandbox instance.
	- This change was made to redirect users who reset their password directly to their Fleet Sandbox Instance.
- `logout.js`: updated to redirect users to the fleetdm.com homepage when they log out.
- `download-sitemap`: `/get-started` changed to `fleetctl-preview`
- `unauthorized.js`: updated to redirect to `/login`
- `bootstrap.js`: updated the admin user in the fake data to have a fake (and expired) Fleet Sandbox
- `routes.js`
	- Changed the current "Get started" page to be at `/fleetctl-preview`
	- Added routes for new pages
	- Added a redirect for `/get-started` to go to the sandbox registration page

New files added:

- `/try-fleet/register`: A page where users can signup to use a Fleet Sandbox instance.
	- `view-register`: This controller redirects users to /try-fleet/sandbox if they are logged in
	- `register.page.js`:
		- This page script uses a handle submitting function to add a `firstName`, `lastName` and `signupReason` to the formData that is submitted to signup.js
		- Once a form is successfully submitted, the user is redirected to /try-fleet/sandbox
- `view-sandbox-teleporter-or-redirect-because-expired`
	- This controller handles the route for `/try-fleet/sandbox-teleporter`
	- Users who have an expired Sandbox instance will be redirected to `/try-fleet/sandbox-expired`
	- If a user does not have a `fleetSandboxURL`, `fleetSandboxDemoKey`, or a `fleetSandboxExpiresAt`. A 500 error will be thrown.
- `try-fleet/sandbox-teleporter`: This page submits a form to the user's Fleet Sandbox Instance's `/demologin` endpoint.
- `/try-fleet/login`: A page for users to log in to their Fleet Sandbox instance 
	- `view-sandbox-login.js`: redirects logged in users to `/try-fleet/sandbox`
	- `sandbox-login.page.js`: redirects users to `/try-fleet/sandbox` after a successful login form submission
- `/try-fleet/new-sandbox`: A page where users who have an existing fleetdm.com account can provision a Fleet Sandbox instance.
	- `view-new-sandbox`: This controller redirects users to `/try-fleet/sandbox` if they have a `fleetSandboxURL`, otherwise, it will take users to `/try-fleet/new-sandbox`
	- `new-sandbox.page.js`: This page script calls the `provision-fleet-sandbox-for-existing-user` action when a user clicks on a "Start Fleet Sandbox" button, then redirects them to `/try-fleet/sandbox` once the action is run.
- `/try-fleet/sandbox-expired`: A page that tells user's that their sandbox instance has expired, and gives them the option to schedule a demo, or join the osquery Slack
- `layout-sandbox.ejs`: A new layout for sandbox pages.




*Original PR description is commented out below, click edit to view*

<!--
This is the current plan for the backend code and page scripts for https://github.com/fleetdm/fleet/issues/5036

 Changes:
  - Updated the User model
     - Added two new attributes: 
        - `fleetSandboxURL`
            - The URL of the user's Fleet sandbox instance
            - Type: string
        - `fleetSandboxExpiresAt` 
           - A JS timestamp of the time when the user's Fleet sandbox instance will expire
           - Type: number
        - 
 
  - Added a new helper: `provision-fleet-sandbox`.
     - This helper will send a POST request to the Cloud provisioned API to request a new Fleet sandbox instance, It then updates the user's record to have a `fleetSandboxURL` and a `fleetSandboxExpiresAt`.
    
   - Added a new action - `/try-fleet/provision-sandbox-and-redirect`
      - This action handles calls the provision-fleet-sandbox helper to provision a new Fleet Sandbox instance and polls the `/healthz` endpoint of the provisioned sandbox until it returns a 200 response. When the Fleet Sandbox instance is live, this action returns the URL of the Fleet Sandbox instance.
      - This action is called when a logged-in user who does not have a Fleet sandbox instance navigates to the sandbox page and clicks a "Start Fleet Sandbox" button
     
   - Added a new action - `/try-fleet/get-sandbox-status` 
      - This action continuously polls the `/healthz` endpoint of the user's Fleet Sandbox instance until it receives a 200 response, it then returns the URL of the live Fleet Sandbox instance
      - This action is called from two places when a user who has provisioned a fleet sandbox instance that is not live navigates to the sandbox page and clicks a "Start Fleet Sandbox" button
     
   - Added a new page - `/try-fleet/register`
      - This page has a form where users can sign up for a Fleet sandbox instance, once they submit the form and their account is created, `get-sandbox-status` is called and the user will be redirected to their Fleet Sandbox instance when it is live
      - This is done in two parts, when a user submits the form, a Fleet Sandbox instance is provisioned when their account is created. After successful form submission, another function runs that calls the `get-sandbox-status` action to poll the `/healthz` endpoint
     
   - Added a new page - `/try-fleet/login`
     - This page is for sandbox users to log in and access their Fleet Sandbox instance.
     - If a license dispenser customer logs in on this page, they will be taken to the Fleet Sandbox page
    
   - Added a new page - `/try-fleet/sandbox`
      - This page's controller checks if a user has a Fleet Sandbox instance, and either:
         - If the user doesn't have a Fleet sandbox instance, they are shown a page where they can click a "start fleet sandbox" button that calls `/try-fleet/provision-fleet-sandbox-and-redirect`, once their Sandbox instance is ready, they will be redirected to it
          - If a user has a Fleet sandbox instance, It will send a single GET request to the `/healthz` endpoint of the Fleet Sandbox instance.
            - If the request was successful, they are redirected to their Fleet Sandbox
            - If the request failed, they are shown a page where they can click a "Start Fleet Sandbox" button that calls `/try-fleet/get-sandbox-status` and they will be redirected to their Fleet Sandbox Instance when the `/healthz` endpoint returns a 200 response.
         - If a user has an expired Fleet Sandbox instance, they are shown a page that says their sandbox expired.
        
- Updated `/entrance/signup.js`
   - Added one new optional input: `sandboxExpirationTimestamp`
   - If a `sandboxExpirationTimestamp` was provided when a user signs up, the `provision-fleet-sandbox` helper is called and a Fleet Sandbox Instance is provisioned for the new user. -->

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -37,6 +37,8 @@ the account verification message.)`,
     organization: {
       required: true,
+      // required: false, //« Change organization to be required: false.
+      // defaultsTo: 'N/A', //« Add a default value, using 'N/A' as a default because this is required on the user model.
```
I realize this is a change in the requirements, but by convention, I think empty string would make sense here.

FYI @noahtalerman - if there’s a reason not to, let Eric know?  Otherwise let’s go with empty string.

eashaw replied:
The organization is required on the User model, so we can't set it to an empty string.

mikermcneil replied:
We're about to make it no-longer required though, right?

mikermcneil replied:
i.e. if you remove `required:true`, the implicit default behavior behind the scenes is `required:false` and `defaultsTo: ''` (which can both be omitted, since they're the default)

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -57,6 +59,13 @@ the account verification message.)`,
+    //   type: 'string',
+    //   description: 'An ISO 8601 timestamp of when this user\'s Fleet sandbox will expire'
+    // },
+
```
The other dates/times we have so far are JS timestamps right?  If so, then I think this should be the same (ie a number), and be named `sandboxExpiresAt` in terms of how it is persisted in the model.  This can be different from the API.

eashaw replied:
Makes sense!

mikermcneil reviewed (changes requested):
More to come later today

mikermcneil reviewed (changes requested):
Nice start @eashaw!

1. Please be sure to read through all of my review comments here before making changes  (some of my suggestions have interconnected explanations)
2. If you think I may be wrong or misunderstanding something given what you're seeing in the wireframes, or if you have another vision you think works better, would you let me know so we can talk it over ASAP?  Totally possible I'm missing things.  Especially re: polling (i.e. if you're seeing something in the wireframes or elsewhere that makes you think I might be confused.)

eashaw commented on website/api/controllers/entrance/signup.js:
```diff
@@ -115,6 +130,12 @@ the account verification message.)`,
       });
     }
 
+    // If a sandboxExpirationTimestamp was provided, this is a user signing up to try Fleet Sandbox and we'll call the provision-fleet-sandbox helper.
```
That makes sense, I've updated the inputs of signup.js and added this attribute to the User model.

eashaw commented:
@mikermcneil I've updated this PR with these changes:

- `view-sandbox`
   - renamed to `view-sandbox-or-redirect`

`Get-sandbox-status`:
- Removed this action

`provision-fleet-sandbox-and-redirect`: 
- Renamed to `provision-fleet-sandbox-for-existing-user`
- This action is called from `/try-fleet/sandbox` when an existing user navigates to the sandbox page but does not have a Fleet Sandbox instance
- Updated to return the fleetSandboxDemoKey with the fleetSandboxURL

`/try-fleet/login`, `login.page.js`, `login.less`, `view-login`
- Renamed to sandbox-login

`User.js`
- Changed organization to be an optional attribute.
- Updated fleetSandboxDemoKey example & description

`signup.js`
- Removed sandboxExpirationTimestamp input
- Added signupReason input
- Removed helper usage, the cloud provisioner request is now handled inline
- This action now polls the `/healthz` endpoint of the provisioned Fleet Sandbox instance until it returns a 200 response.

eashaw commented on website/api/controllers/entrance/signup.js:
```diff
@@ -115,6 +130,12 @@ the account verification message.)`,
+    // If a sandboxExpirationTimestamp was provided, this is a user signing up to try Fleet Sandbox and we'll call the provision-fleet-sandbox helper.
+    ///  this helper will make a POST request to the cloud provisioner, add the URL of the fleet sandbox instance and the expiration timestamp to the user's record
+    // if(sandboxExpirationTimestamp) {
+        // await sails.helpers.provisionFleetSandbox.with({userId: newUserRecord.id, sandboxExpiresAt: sandboxExpirationTimestamp})
```
I've removed the helper and added the code inline.

mikermcneil commented:
@eashaw Great, thanks!  Let me know when you're ready for final code review.

eashaw commented:
@mikermcneil this PR is ready for the final code review.

This is the updated description:

Changes to existing code:
---
- `models/User.js`: 
	- Added two new optional attributes:
		- `fleetSandboxURL`: `type: 'string'` The URL of the user's Fleet Sandbox instance
		- `fleetSandboxExpiresAt`: `type: 'number'` A JS Timestamp representing when the user's Fleet Sandbox instance will expire
	- Changed the `organization` attribute to be optional

- `signup.js`: 
	- Added a new exit: `couldNotProvisionSandbox`
	- Changed the `organization` input to be optional, defaulting to a blank string.
	- Added `signupReason` as an optional input, with a default value of 'Buy a license'
	- If `signupReason` is 'Try Fleet Sandbox', this action will:
		- Send a POST request to the cloud provisioner API
		- Updates the user's record with a `fleetSandboxURL` and `fleetSandboxExpiresAt`
		- Polls the new Fleet Sandbox instance's `/healthz` endpoint until it returns a `200` response.

- `logout.js`: Updated to redirect users to the Fleet homepage after logging out

- `update-password.js`: This action has been updated to change a user's password on their Fleet Sandbox Instance
	- Added a new exit: `couldNotChangeSandboxPassword`
	- If a user has a `fleetSandboxURL`, and their instance has not expired, the action makes two requests to the Fleet Sandbox instance:
		- A POST request to the `/login` endpoint of the Sandbox instance
		- A POST request to the `/change_password` endpoint of the Sandbox instance

- `send-password-recovery-email.js`:
	- Added a new optional input `passwordResetRequestedFrom`, this defaults to `customers` if not provided.
	- Updated this action to send the url of the new password page. either `/customers/new-password` (For license dispenser customers), or `/try-fleet/new-password` (for Fleet Sandbox users) as a variable called `passwordResetLocation`

- `update-password-and-login.js`: 
	- Added a new exit: `couldNotChangeSandboxPassword`
	- If a user has a `fleetSandboxURL`, and their instance has not expired, the action makes two requests to the Fleet Sandbox instance:
		- A POST request to the `/login` endpoint of the Sandbox instance
		- A POST request to the `/change_password` endpoint of the Sandbox instance

- `forgot-password.page.js`:
	- If a user comes to this page from the `/try-fleet/login`, a `passwordResetRequestedFrom` value is added to the formData that is submitted to `send-password-recovery-email.js` and will change the password reset link to take Fleet Sandbox users to `/try-fleet/new-password` instead of `/customers/new-password`
- `new-password.page.js`: 
	- Updated the password requirements in the formRules, New passwords need to be at least 12 characters, and must contain a letter, number and special character
	- If a user came to this page from the `/try-fleet/new-password` route, they will be redirected to `/try-fleet/sandbox` where they will be logged in to their Fleet sandbox instance

- `email-reset-password.ejs`: Changed the password reset url to use the `passwordResetLocation` value send from `send-password-recovery-email.js`

- `forgot-password.ejs`: Updated the "Back to login" login to send users to `/try-fleet/login` if a `redirectToSandbox` value in the page locals

- `routes.js`:
	- replaced the route for `/get-started` with a redirect to `/try-fleet/register`
	- added routes for new pages & action

- `entrance/signup.ejs`: Added a form error for new custom password formRule
- `entrance/signup.page.js`: Added a new custom formRule for checking that a password meets our requirements (12 characters and includes at least one letter, number, and special character)
- `customers/dashboard.ejs`: Added a form error for new custom password formRule
- `customers/dashboard.page.js`: Added a custom formRule for password requirements

Additions:
---
- Created a new action: `provision-fleet-sandbox-for-existing-user`:
	- This action takes a userID as an input and is called from the `/try-fleet/new-sandbox` page
	- This action has three exits: `success`, `userHasExistingSandbox`, and `couldNotProvisionSandbox`
	- If a user does not have a Fleet Sandbox Instance, this action will:
		- Request a new sandbox instance from the cloud provisioner API
		- Updates the users record with `fleetSandboxURL` and `fleetSandboxExpiresAt` values
		- Polls the new Fleet Sandbox instance's `/healthz` endpoint until it returns a `200` response.
	- If a User has a Fleet Sandbox instance, or either of the requests fail, then this will return an error.

- `/try-fleet/register`: A page where users can signup to use a Fleet Sandbox instance.
	- `view-register`: This controller redirects users to `/try-fleet/sandbox` if they are logged in
	- `register.page.js`: 
		- This page script uses a handle submitting function to add a `firstName`, `lastName` and `signupReason` to the formData that is submitted to `signup.js`
		- Once a form is successfully submitted, the user is redirected to `/try-fleet/sandbox`

- `view-sandbox-or-redirect`
	- This controller handles the route for `/try-fleet/sandbox`
	- If a user does not have a sandbox instance, they are redirected to `/try-fleet/new-sandbox`
	- Users who have an expired Sandbox instance will be redirected to `/try-fleet/sandbox-expired`
	- If a user has a unexpired Fleet Sandbox instance this controller will poll the `/healthz` endpoint of their Fleet Sandbox until it returns a 200 response, and take users to the sandbox page, where a form submission will log them into their Fleet Sandbox instance.

- `/try-fleet/login`: A page for users to log in to their Fleet Sandbox instance 
	- `sandbox-login.page.js`: redirects users to `/try-fleet/sandbox` after a successful login form submission
	-  `view-sandbox-login.js`: redirects logged in users to `/try-fleet/sandbox`

- `/try-fleet/sandbox`: This page logs users into theyre sandbox instance and redirects them to their Fleet Sandbox dashboard.
	- `sandbox.page.js`: submits a form to the `/demologin` endpoint of the user's Fleet Sandbox instance, after the form is submitted, they are redirected to their Fleet Sandbox dashboard.

- `/try-fleet/new-sandbox`: A page where users who have an existing fleetdm.com account can provision a Fleet Sandbox instance.
	- `view-new-sandbox`: This controller redirects users to `/try-fleet/sandbox` if they have a `fleetSandboxURL`, otherwise it will take users to `/try-fleet/new-sandbox`
	- `new-sandbox.page.js`: This page script calls the `provision-fleet-sandbox-for-existing-user` action when a user clicks on a "Start Fleet Sandbox" button, then redirects them to `/try-fleet/sandbox` once the action is run.

- `/try-fleet/sandbox-expired`: A page that tells user's that their sandbox instance has expired.

- `layout-sandbox.ejs`: A new layout for sandbox pages.

mikermcneil commented:
@eashaw Please revisit my [requested changes](https://github.com/fleetdm/fleet/pull/6380/files/1e1d2a47142490ca83a9629bb271d8907be90380).  It looks like some were not applied (e.g. the page rename to better match the purpose of the "sandbox" page).

If that's intentional, can you let me know where I'm wrong?

mikermcneil commented on website/api/models/User.js:
```diff
@@ -66,7 +66,7 @@ module.exports = {
     organization: {
       type: 'string',
-      required: true,
+      required: false,
```
`required:false` is always the default for model attributes, so it can be omitted.

eashaw replied:
I've removed that line

mikermcneil commented on website/config/routes.js:
```diff
@@ -188,7 +179,60 @@ module.exports.routes = {
+      layout: 'layouts/layout-customer',
+      pageTitleForMeta: 'New password | Fleet for osquery',
+      pageDescriptionForMeta: 'Change the password for your Fleet account.',
+      redirectToSandbox: true,
```
It would be more consistent to send down the `redirectToSandbox` local flag from within the view action, rather than in the routes file.  (Otherwise it's hard to know where the page script is getting it from.  And this is the consistent approach that matches how existing dynamic locals like `token` work.)

Same thing for the other places.

It's fine to leave layout, and the page title and description in the routes file.

eashaw replied:
That makes sense, I updated the view actions and routes for these pages.

mikermcneil reviewed (changes requested):
A couple of initial points.  I'll review again ASAP when I have a better understanding of what's going on with some of the requested changes from last time that didn't make it in.

mikermcneil commented on website/api/controllers/try-fleet/provision-fleet-sandbox-and-redirect.js:
```diff
@@ -0,0 +1,44 @@
+  },
+
+
+  fn: async function (inputs) {
```
@eashaw Why was this action not removed?

mikermcneil commented on website/config/routes.js:
```diff
@@ -26,15 +26,6 @@ module.exports.routes = {
-      pageDescriptionForMeta: 'Learn about getting started with Fleet.'
-    }
-  },
-
```
To maintain backwards compatibility of links, let's replace this with a redirect down in the redirects section.  (To where?  Homepage is probably the best bet.)

mikermcneil replied:
@eashaw Nope, all good!  I just missed the redirect somehow.

mikermcneil reviewed (changes requested):
One other unrelated edge case that jumped out for backwards compatibility of links.

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -115,6 +131,43 @@ the account verification message.)`,
+        'sandbox_expiration': new Date(fleetSandboxExpiresAt).toISOString(), // sending expiration_timestamp as an ISO string.
+      })
+      .timeout(5000)
+      .intercept('non200Response', 'couldNotProvisionSandbox');
```
I'm also not seeing the shared secret here, or the strategy involving the `demoKey` as specced at the bottom of https://docs.google.com/document/d/1DUR6XtsuxTrXIrpkRR5J1KFfADwL9CrcSRyR2joozZg/edit

eashaw replied:
The shared secret has been added to the cloud provisioner request as a header using an environment variable named `cloudProvisionerSecret`

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -115,6 +131,43 @@ the account verification message.)`,
+          if(serverResponse && serverResponse.statusCode) {
+            return serverResponse.statusCode === 200;
+          }
+        });
```
@eashaw Please set up the other argument that limits how many times this can run, or how long (I forget, check getDef for how to use this)

eashaw replied:
I added a 10-second timeout to `sails.helpers.flow.until()`

eashaw commented:
@mikermcneil I've updated this PR with the requested changes. 

There was a change that is different from what we discussed this morning for existing fleetdm.com users. I've updated the login action to provision a Fleet Sandbox instance when a user logs in. If this isn't OK please let me know and I will revert that change ASAP.

noahtalerman commented:
@mike-j-thomas, @mikermcneil heads up, on the `/try-fleet/sandbox` expired page, we'd like to direct users to "Schedule a demo" and navigate them to the following link: https://calendly.com/fleetdm/demo.

We also would like to remove the "To continue using Fleet..." subheader.

I just updated the following screen in the fleetdm.com Figma: https://www.figma.com/file/yLP0vJ8Ms4GbCoofLwptwS/%E2%9C%85-fleetdm.com-(current%2C-dev-ready)?node-id=4043%3A15202

![Screen Shot 2022-07-22 at 12 16 38 PM](https://user-images.githubusercontent.com/47070608/180481234-8e881c3a-8da6-4281-82bf-f41a4b9b615a.png)

This is instead of pointing the user to the deployment docs.

mikermcneil commented on website/api/controllers/account/update-password.js:
```diff
@@ -42,6 +46,37 @@ module.exports = {
+        }
+      }
+    }
+
```
My understanding is that their fleetdm.com and password is the only way users will ever access Fleet Sandbox, so changes to that password don't need to be sent to the Fleet instance.

If there was a change to that requirement or you think there could be confusion, could you please let @noahtalerman know ASAP?

eashaw replied:
@mikermcneil The Fleet Sandbox instance needs the user's email address and (hashed) password to log the user in with the `/demologin` endpoint. 
If the user changes their password on fleetdm.com, they wouldn't be able to log in to their Fleet Sandbox instance with the credentials that are sent to the `/demologin` endpoint.

mikermcneil replied:
Ah, I see.  So this is another side effect of the change to the authentication strategy that was made earlier this month.  The user would never actually know they are using those credentials to sign in, it’s just that fleetdm.com is using those credentials behind the scenes instead of a demoKey.

@zhumo FYI this means it’ll be important to hide the “Change my password” functionality in the Fleet product, when it is running as part of Fleet Sandbox.  I don’t remember seeing it in the wireframes, because it didn’t matter previously before the underlying authentication mechanism was changed a week or so ago.

eashaw replied:
@mikermcneil I realized when I was typing that there was a different/possible better way to do this.

How does this sound:

- When signing up, a `fleetSandboxDemoKey` string is created, using a combination of `sails.helpers.strings.random('url-friendly', 12)`, The new user's email address, and the sandbox expiration JS timestamp. It uses a combination of those three values so that we can make sure that every `fleetSandboxDemoKey` we send to the Cloud Provisioner API meets the Fleet password requirements
- The `fleetSandboxDemoKey` is used in place of the User's (hashed) password when provisioning a Fleet Sandbox Instance
- This value is stored on the User model and is only used to log the user into their Fleet Sandbox instance, which will only be valid for 24 hours.

If I make that change, we can remove this portion of the code **and** we won't need update our password requirements on fleetdm.com (which we probably don't need to do regardless, because we using hashed versions of the users password to create their Fleet Sandbox user)

zhumo replied:
My understanding of Eric's proposal is that we would have the user log in to Fleetdm.com, which will retrieve this `fleetSandboxDemoKey`, and secretly in the back end, we are using that key to authenticate them into fleet sandbox. This way, they do not need to maintain two separate credentials? 

If that's the correct understanding, then that sounds good to me from a product perspective.

mikermcneil replied:
@eashaw I like your thinking, in terms of not sending a real hashed password to avoid confusion.  What would happen if the email is changed on fleetdm.com?  Does email need to be included in the demo key or can it be excluded and some other special character and number included instead?  Is there any reason to include the expiry in this other than to make sure there’s a number?

@zhumo either way, disabling changing one’s own email and password in the Fleet product for sandbox is probably still a good idea, because otherwise users might misunderstand. Also if they are able to change the password, they could still break their ability to access the sandbox.

eashaw replied:
>Does email need to be included in the demo key or can it be excluded and some other special character and number included instead? Is there any reason to include the expiry in this other than to make sure there’s a number?

Those would only for the guaranteed number and special character. We could also just add an arbitrary number and special character to the random string.

eashaw replied:
@mikermcneil I opted to use `sails.helpers.strings.uuid()` to generate the `fleetSandboxDemoKey` because it meets all the password requirements of the Sandbox Instance without being changed.

zhumo replied:
@mikermcneil I think it's already been hidden per: https://github.com/fleetdm/fleet/issues/6544.

The other place where this might happen is if you do forgot pw. If we are using fleetdm.com's credentials in place of fleet sandbox's credentials, then all forgot pw stuff should be handled via fleetdm.com not sandbox. 

@eashaw , I see that you are assigned to this one: https://github.com/fleetdm/fleet/issues/5036. Will this handle the forgot password flow in a way such that a user can't reset their sandbox pw?

eashaw replied:
@zhumo Yes! The forgot password flow will only handle fleetdm.com credentials. The password for the Fleet Sandbox instance would be stored as a separate attribute on the User model and can't be reset.

mikermcneil commented on website/api/controllers/entrance/send-password-recovery-email.js:
```diff
@@ -55,7 +65,7 @@ module.exports = {
-        firstName: userRecord.firstName,
+        passwordResetLocation: passwordResetLocation,
         token: token
       }
```
Does this need to exist, or can we use the same password recovery flow for both and leave this code as-is?  I didn't expect to see password recovery change at all as part of this PR, and it was not included in the last set of changes I reviewed, so I'm surprised to see it now.

eashaw replied:
>Does this need to exist, or can we use the same password recovery flow for both and leave this code as-is?

The password recovery flow that was planned for Fleet Sandbox in #5036 has us redirect users to their logged-in Fleet Sandbox after they successfully changed their password.

>I didn't expect to see password recovery change at all as part of this PR, and it was not included in the last set of changes I reviewed, so I'm surprised to see it now.

This is my fault, when I was drafting this PR with code comments, I was going off our (now changed) authentication plan for Fleet Sandbox, and the password recovery flow would only need a redirect (which I had forgotten to include in the draft plan). 
When I started writing the code for this PR, I updated this flow because the requirements changed to POST a login form with the user's credentials to the Sandbox Instance instead of using a generated key stored on the User model.

mikermcneil replied:
@eashaw Got it, very helpful, thank you.

Instead of adding more pages and actions which have to be maintained, would it work to use one password recovery flow and have it look to see if sandbox is not expired, and if not, redirect to it, else redirect to the license dispenser portal?

eashaw replied:
>Instead of adding more pages and actions which have to be maintained, would it work to use one password recovery flow and have it look to see if sandbox is not expired, and if not, redirect to it, else redirect to the license dispenser portal?

That would work and it would be an easy change. My only concern with this approach is that Fleet Sandbox users with expired instances would be redirected to the [`/customers/new-license` page](https://www.figma.com/file/Jzo81K6E4jC0mcjD4JsWM8/%F0%9F%9A%A7-fleetdm.com-(scratchpad)?node-id=1021%3A2046) with no explanation why. These users would have to navigate back to the Fleet homepage and then click the "try it now" button to see that their sandbox is expired.

mikermcneil replied:
Ah, I see.  Could we look to see if they have a credit card saved and only link them to license dispenser in that case, otherwise show them the expired sandbox page?

eashaw replied:
Yes! I added the redirect to [`view-login.js`](https://github.com/fleetdm/fleet/pull/6380/commits/2e8e26b47e5d6ccfe80e96ad2e48f24dabdc2601#diff-4deb19da69982f6ce9c1c5616ae0abae51dffb1e3702c8182306e68c15a49139)

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -36,7 +36,8 @@ the account verification message.)`,
 
     organization: {
-      required: true,
+      required: false, //« Change organization to be required: false.
```
Please see https://github.com/fleetdm/fleet/pull/6380#discussion_r909175747.

mikermcneil replied:
And https://github.com/fleetdm/fleet/pull/6380#discussion_r924965894

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -55,6 +56,14 @@ the account verification message.)`,
+
+    // Add optional input
+    signupReason: {
+      required: false,
```
Please see https://github.com/fleetdm/fleet/pull/6380#discussion_r909175747

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -36,7 +36,8 @@ the account verification message.)`,
     organization: {
-      required: true,
+      required: false, //« Change organization to be required: false.
+      defaultsTo: '', //« Add a default value
```
Please remove the comments talking about changes made in this PR.

mikermcneil reviewed (changes requested):
Hey @eashaw, I'm still seeing some changes that didn't make it in from the last time I reviewed.  Can you please revisit these and then let me know when it is ready for review again?

I left a few comments about changes that can be left out as well (or if you think I'm wrong, please let @noahtalerman know ASAP.)

eashaw commented on website/api/controllers/account/update-password.js:
```diff
@@ -30,6 +30,10 @@ module.exports = {
+    },
+
+    couldNotChangeSandboxPassword: {
+      description: 'An error occurred while changing this user\'s password on their Fleet sandbox instance',
```
Makes sense! The update password changes have been removed from this PR, but I've updated the exits and errors in [`signup.js`](https://github.com/fleetdm/fleet/pull/6380/files#diff-9c6a73c96fdd057f48cfb92925ac8cbf02462ac4889c1ee6dd5bc88cec546465) to use 500 errors.

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -36,7 +36,7 @@ the account verification message.)`,
 
     organization: {
-      required: true,
+      defaultsTo: '',
```
`defaultsTo: ''` is the implicit default value for `type: 'string'`, so this can be removed.

mikermcneil replied:
I'll tweak this in a commit to save a roundtrip here.

mikermcneil replied:
Actually @eashaw, my reasoning for the above is faulty.

While we can still remove `defaultsTo: ''` here, the reason is more nuanced.  While you never need `required: false`, no matter what, specifically in the case of validating inputs, the value will be `undefined` for an optional input that ends up being unspecified at runtime, regardless of its type.

The reason why this `defaultsTo: ''` can be omitted here is that `organization` is passed through to `User.create()`, where `organization` is an optional attribute.

So by passing in `undefined`, it leaves it up to the ORM, which then, because of how `type:'string'` works in Waterline, is actually persisted as empty string (`''`) automatically.  And that's what will come back any time it is queried.

Hope that helps- sorry for any confusion my incomplete explanation above might have caused.

mikermcneil replied:
See https://github.com/fleetdm/fleet/pull/6380/commits/2665abef1dbb471086cc5888be01f805ef456472#diff-9c6a73c96fdd057f48cfb92925ac8cbf02462ac4889c1ee6dd5bc88cec546465L39 and https://github.com/fleetdm/fleet/pull/6380/commits/e150047bb9adc2ad237ae02994ef3ad0528ed231

mikermcneil commented on website/assets/js/pages/try-fleet/register.page.js:
```diff
@@ -0,0 +1,58 @@
+        // Custom password validation to ensure it contains at least one letter, one number, and one special character. TODO: full list of special characters
+        custom: (password)=>{
+          return !! password.match(/[\!\@\#\$\%\^\&\*]/) && password.match(/\d/) && password.match(/\w/);
+        }
```
Technically we don't need these stricter validations anymore, so we could revert it out save it for later, if needed. (This would avoid people who have already signed up from potentially having to change their password to make it valid.)

mikermcneil commented on website/assets/js/pages/try-fleet/register.page.js:
```diff
@@ -0,0 +1,58 @@
+      argins.firstName = argins.emailAddress.split('@')[0];
+      argins.lastName = argins.emailAddress.split('@')[1];
+      argins.signupReason = 'Try Fleet Sandbox';
+      await Cloud.signup.with(argins);
```
Based on how handleSubmitting functions are designed to work, this should be `return await Cloud.signup.with(argins);`   (it likely won't matter in this case, but it's a good idea to do this everywhere, so that you can get ahold of the server response data easily in your HTML when/if needed down the road, without having to come back in here.  Plus it's the convention)

mikermcneil commented on website/views/emails/email-reset-password.ejs:
```diff
@@ -3,6 +3,6 @@
 <p style="margin-bottom: 32px;">Someone requested a password reset for your Fleet account.<br>Please click below to reset your password:</p>
 
   <div style="background: #6A67FE; display: inline-block; color: white; text-decoration: none; border-radius: 4px; text-align: center; width: 151px; padding: 8px 16px 8px 16px" >
-    <a style="margin: auto; font-size: 16px; text-decoration: none; color: white; line-height: 22px;" href="<%= url.resolve(sails.config.custom.baseUrl,'/customers/new-password')+'?token='+encodeURIComponent(token) %>">Update password</a>
```
I agree with this change for the sake of readability, but since this is just adding a space character, I'd omit this change from this PR to avoid adding this to the `git blame` and making us think emails were changed in this PR (just since there's been a lot of back and forth and we'll probably kinda forget)

mikermcneil commented on website/views/pages/try-fleet/register.ejs:
```diff
@@ -0,0 +1,42 @@
+          <div class="form-group mb-3">
+            <input id="password" type="password" class="form-control d-flex w-100" :class="[formErrors.password ? 'is-invalid' : '']" v-model.trim="formData.password" placeholder="Choose a password">
+            <div class="invalid-feedback mt-2" v-if="formErrors.password === 'minLength'">Password too short.</div>
+            <div class="invalid-feedback mt-2" v-if="formErrors.password === 'custom'">Passwords must be at least 12 characters long, and must contain at least one letter, number and special character (!@#$%^&*).</div>
```
If you get rid of the validation in the other places, I'd suggest removing it here, even though painful.

Also, the & should be &amp; so that it is an encoded HTML entity. It might technically work as-is, but the red highlight you see is because ampersands are supposed to be escaped as &amp; (aka an HTML entity) in HTML text nodes (aka loose text that lives between <foo></foo>)

mikermcneil commented on website/views/pages/try-fleet/sandbox-login.ejs:
```diff
@@ -0,0 +1,44 @@
+            <input id="password" type="password" class="form-control d-flex w-100" :class="[formErrors.password ? 'is-invalid' : '']" v-model.trim="formData.password" placeholder="Password">
+            <div class="invalid-feedback mt-2" v-if="formErrors.password === 'minLength'">Password too short.</div>
+
+            <div class="invalid-feedback mt-2" v-if="formErrors.password === 'custom'">Passwords must be at least 12 characters long, and must contain at least one letter, number and special character (!@#$%^&*).</div>
```
I don't think you actually need anything other than "Please enter a password" here.  Like, if you enter a password that's too short, then it won't match.

Also if you leave out the validation here, then if password restrictions get gnarlier, people with existing passwords can still log in.

This would match what you're doing in the corresponding page script.  (i.e. neither of the "password is too short" or "password must be at least 12" messages would ever show, because neither validation is actually applied in the page script.)

mikermcneil commented on website/api/controllers/try-fleet/view-sandbox-or-redirect.js:
```diff
@@ -0,0 +1,50 @@
+  friendlyName: 'View sandbox',
+
+
+  description: 'Display "Sandbox" page or redirect users to their Fleet Sandbox instance.',
```
This friendly name, filename, and description could use some help.  Because of the nature of the login trick, being extra explicit here will be helpful.  Especially since the form post kinda "feels" like a redirect.  Also, this is not either displaying a page or redirecting users to their Fleet Sandbox instance:  it's either (A) displaying a brief interstitial form post that autosubmits to a subdomain and returns HTML which then sets browser localstorage to contain credentials and consider the user logged in, or (B) redirecting because the user's Fleet Sandbox instance is expired.

So how about:

- filename: `view-sandbox-teleporter-or-redirect-because-expired.js`
- `friendlyName: 'View sandbox teleporter or redirect because expired'`
- `description: 'Display "Sandbox teleporter" page (an auto-submitting interstitial HTML form used as a hack to grab a bit of HTML from the Fleet Sandbox instance, which sets browser localstorage to consider this user logged in and "teleports" them, magically authenticated, into their Fleet Sandbox instance running on a different domain), or redirect the user to a page about their sandbox instance being expired.'`
- `moreInfoUrl: 'https://github.com/fleetdm/fleet/pull/6380'`

mikermcneil commented on website/api/controllers/try-fleet/view-sandbox-or-redirect.js:
```diff
@@ -0,0 +1,50 @@
+  exits: {
+
+    success: {
+      viewTemplatePath: 'pages/try-fleet/log-into-fleet-sandbox-and-redirect',
```
`'pages/try-fleet/sandbox-teleporter'`

(good exit description btw)

mikermcneil commented on website/assets/js/pages/try-fleet/log-into-fleet-sandbox-and-redirect.page.js:
```diff
@@ -0,0 +1,27 @@
+parasails.registerPage('log-into-fleet-sandbox-and-redirect', {
```
Because I didn't understand the purpose of this page before, due to the API for the Fleet provisioner changing prior to my last review, I was misleading here.

Since, based on my other suggested edits in this review, the view action for this page becomes `view-sandbox-teleporter-or-redirect-because-expired`, by convention, the name of this page script, the corresponding LESS file, and the corresponding EJS file should all be `sandbox-teleporter.page.js`, `sandbox-teleporter.less`, and `sandbox-teleporter.ejs` respectively.

Make sense?

mikermcneil commented on website/assets/js/pages/try-fleet/log-into-fleet-sandbox-and-redirect.page.js:
```diff
@@ -0,0 +1,27 @@
+    //…
+  },
+  mounted: async function() {
+    document.forms['demologin'].submit();
```
I would add a comment here like:

```js
// Confused?  Understandable, this approach is a bit unusual.  See this page's view action for more info on what this code is doing and why, as well as a link where you can read more information.
```

mikermcneil commented on website/assets/styles/pages/try-fleet/sandbox.less:
```diff
@@ -0,0 +1,5 @@
+#sandbox {
```
This file can be deleted.  Please see my review comment in the corresponding page script.

mikermcneil commented on website/views/pages/try-fleet/log-into-fleet-sandbox-and-redirect.ejs:
```diff
@@ -0,0 +1,8 @@
+<div id="log-into-fleet-sandbox-and-redirect" v-cloak>
+  <form method="post" action="<%=me.fleetSandboxURL%>/api/latest/fleet/demologin" id="demologin">
```
I would add a comment here (inside the `<div>` otherwise it could cause Vue.js issues!!), like:

```
<!-- Confused?  Understandable, this approach is a bit unusual. See this page's view action for more info on what this code is doing and why, as well as a link where you can read more information. -->
```

mikermcneil commented on website/assets/styles/pages/try-fleet/new-sandbox.less:
```diff
@@ -0,0 +1,69 @@
+#new-sandbox {
```
Because I didn't understand the purpose of this page before, due to the API for the Fleet provisioner changing prior to my last review, I was misleading here.

Since, based on my other suggested edits in this review, the view action for this page becomes `view-sandbox-teleporter-or-redirect-because-expired`, by convention, the name of the page script, the corresponding LESS file, and the corresponding EJS file should all be `sandbox-teleporter.page.js`, `sandbox-teleporter.less`, and `sandbox-teleporter.ejs` respectively.

Make sense?

mikermcneil replied:
You'll also want to make sure:
- the DOM id in the page script  (`'sandbox-teleporter'`)
- the DOM id in the HTML (`id="sandbox-teleporter"`)
- the DOM selector in the LESS file (`#sandbox-teleporter`)
- and `importer.less`

all match too.

mikermcneil commented on website/config/routes.js:
```diff
@@ -180,7 +171,33 @@ module.exports.routes = {
+  },
+
+  'GET /try-fleet/sandbox': {
+    action: 'try-fleet/view-sandbox-or-redirect',
```
`action: 'try-fleet/view-sandbox-teleporter-or-redirect-because-expired'`

(See my review comments in the corresponding page script for more info.)

mikermcneil reviewed (changes requested):
@eashaw Great work, Eric.  I left you a lot more requested changes, but this time, I'm confident that, if you read through them all, then go and apply them one by one without pushing, you'll be in great shape and this will be ready to go.

I'd like to review one more time to make sure we ended up on the same page in our understanding, so I'll plan on checking out the actual commits you make rather than re-reviewing all files changed.

Thanks!

eashaw commented on website/config/routes.js:
```diff
@@ -26,15 +26,6 @@ module.exports.routes = {
-      pageTitleForMeta: 'Get started | Fleet for osquery',
-      pageDescriptionForMeta: 'Learn about getting started with Fleet.'
-    }
-  },
```
I went with option 2 and added a route for the original get-started page

eashaw commented:
@mikermcneil I've made the requested changes and I've updated the changes in the PR description.

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -100,46 +104,46 @@ the account verification message.)`,
-        The newly created User record has been deleted, and the user will be asked to try signing up again.
-        Here is the response from the Cloud Provisioner API: ${cloudProvisionerResponse}`
+        `When provisioning a Fleet Sandbox instance for this new user, the response data from the cloud provisioner API was malformed. It did not contain a valid Fleet Sandbox instance URL in its expected "URL" property.
+        The newly created User record , and the user will be asked to try signing up again.
```
Looks like some text went missing here

mikermcneil replied:
I don't think it deleted the User record right?  It's just that no User record was created.

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -100,46 +104,46 @@ the account verification message.)`,
-      throw new Error('When attempting to provision a new user\'s Fleet Sandbox instance, the Cloud provisioner gave a non 200 response. The incomplete user record has not been saved in the database, and the user will be asked to try signing up again.');
+    .intercept(['requestFailed', 'non200Response'], (err)=>{
+      // If we recieved a non-200 response from the cloud provisioner API, we'll throw a 500 error.
+      return new Error('When attempting to provision a new user\'s Fleet Sandbox instance, the cloud provisioner gave a non 200 response. The incomplete user record has not been saved in the database, and the user will be asked to try signing up again. Raw response received from provisioner: '+err.stack);
```
I overlooked this before, but similar to the other spot in this file:

> When attempting to provision a new user\'s Fleet Sandbox instance

Which user?

It would be helpful to include the email address of the user here too (similar to https://github.com/fleetdm/fleet/pull/6380#discussion_r929567471)

mikermcneil replied:
In case it saves you time:  (I'll leave it to you to edit so I don't mess you up)

```js
return new Error('When attempting to provision a new Fleet Sandbox instance for a new user who just signed up ('+emailAddress+'), the cloud provisioner gave a non 200 response. The incomplete user record has not been saved in the database, and the user will be asked to try signing up again. Raw response received from provisioner: '+err.stack);
```

mikermcneil reviewed (changes requested):
2 things:
- some text in an error message got garbled
- pointing out another opportunity to provide the email address of the affected user

mike-j-thomas reviewed (commented):
@eashaw and @mikermcneil, I have reviewed this and identified the following UX issues.

As I have not been managing this task, I'll leave it to you guys to prioritize which of these should be addressed prior to launch. Some are "nice to haves," so can be added iteratively after release.

- Because Sandbox is borrowing `/customers/forgot-password`, hitting "Back to login" redirected me to License Dispenser login rather than Sandbox login. 

- Creating an account returned this error:

<img width="482" alt="image" src="https://user-images.githubusercontent.com/78363703/181193460-593c5ddc-f581-45f3-8ef0-7511a6feafb5.png">

along with this output in Terminal:

<img width="1241" alt="image" src="https://user-images.githubusercontent.com/78363703/181198340-69d5841e-b557-4803-9ff2-e687712347c8.png">

- As Sandbox loads, it would be a better experience if this page did not appear. If it is necessary, a blank page with a spinner and message, such as "Reticulating splines," would be preferable.

<img width="867" alt="image" src="https://user-images.githubusercontent.com/78363703/181194648-ae1e9249-82af-491f-b8bc-d78335e21247.png">

- When hitting the back button from Sandbox redirects the user back to the page mentioned above, causing the user to get caught in a loop.

- Logging out is very awkward. I.e., navigating to /customers. A user would not know how to do this. Not important for now, but I feel like users should be able to log out. Maybe they are on a shared computer or something. A potential solution could be that the user is taken to a version of /try it out that tells them how long is remaining on their trial and gives a log out option.

<img width="1167" alt="image" src="https://user-images.githubusercontent.com/78363703/181196663-2bca3213-4fc9-4b25-8991-3882fd676858.png">

- What happens if a user signs up for Sandbox on Mobile? It takes them to a product that looks broken. Should we display a message saying that Fleet is not currently optimized for mobile devices?

<img width="386" alt="image" src="https://user-images.githubusercontent.com/78363703/181197112-f3f16062-d8cd-42f3-a724-fc5c96ed7404.png">

- The placeholder text for changing the password (••••••) confused me. I thought I needed to delete them. These inputs should be empty.

<img width="430" alt="image" src="https://user-images.githubusercontent.com/78363703/181197410-f5ab63f8-2277-4b4f-be23-f6ee5926de5d.png">

- Not a blocker, but I would prefer these errors to reset once I have deleted the bad input. Again, the placeholder •••••• should be removed. It looks like I still have a bad input whereas it's actually an empty input.

<img width="575" alt="image" src="https://user-images.githubusercontent.com/78363703/181197599-a4f3a901-3b04-45d4-9283-74064ff9688b.png">

mikermcneil commented:
Hey @mike-j-thomas, @eashaw, and @zhumo, here are the changes I think we should prioritize 3 or 4 things from the above:

> The placeholder text for changing the password (••••••) confused me. I thought I needed to delete them. These inputs should be empty.

(a) Remove placeholder text

> Creating an account caused this error…along with this output in Terminal

(b) Verify why the underlying error is not being included and then fix it.  To test, ensure that when the scenario that Mike T encountered causing the error happens, the error in the terminal contains enough information to tell what happened, including the underlying error in the call to the provisioner API.

> signup failing

(c) Was this happening consistently?  What was it happening?  Because of the other issue above, there was no good error, so once we fix that, we should make sure that signup works, every time.

> white screen and broken back button during signup

(d) The white screen and broken back button behavior for the signup flow is not a good experience, but it is an unfortunate side effect of the change that was made to the plan for authentication.

@mike-j-thomas @eashaw  Do you think this is shippable with the white screen hanging as long as it does?   We could add a spinner and loading state pretty easily.  Maybe worth it.  Thoughts?

<img width="766" alt="image" src="https://user-images.githubusercontent.com/618009/181304353-8447612a-4dfa-4fdf-b8e2-48f16383a80c.png">


I'll schedule 30m with MIke T and Mo to go through the rest of the points above, and also to see if there's an easy way we can make the default organization name shown in the top left of the screen more friendly, instead of displaying a random string.

mike-j-thomas commented:
@mikermcneil, good call about the random string for org name. I forgot to include that.

> Do you think this is shippable with the white screen hanging as long as it does? We could add a spinner and loading state pretty easily. Maybe worth it. Thoughts?

I think it is shippable providing we add a loading state, and hide the masthead. I'm not sure what can be done about the back button loop?

eashaw commented:
@mike-j-thomas @mikermcneil 
I've updated this PR with these changes:
- Removed the placeholders in the password inputs of the new password page
- Updated the sandbox-teleporter page script to send users navigating back to the sandbox-teleporter page with their browsers back button to the fleetdm.com homepage
- The header is now hidden on the sandbox-teleporter
- Added a loading spinner to sandbox-teleporter
- Added a "Log out" button to the main website header navigation if users are logged in
- Updated errors on signup.js to include the new user's email address
- Added a check to signup.js to see if a user exists with the email provided before the cloud provisioner request is sent

mikermcneil commented:
Awesome, thanks Eric.  

>Added a "Log out" button to the main website header navigation if users are logged in

In case you haven't already, reminder to test this on mobile if you haven't already, since I think our nav may swap out and display:none (i.e. duplicated in the HTML so both spots would need to be tested)

mike-j-thomas reviewed (commented):
@eashaw,

> Removed the placeholders in the password inputs of the new password page

I don't know how to view this page. I can't remember how to get into the Sails console to view grab the link from the command-line output.

> Updated the sandbox-teleporter page script to send users navigating back to the sandbox-teleporter page with their browsers back button to the fleetdm.com homepage

I had to hit the back button twice. The first time teleported back into Sandbox. The second time took me to the homepage.

> Added a loading spinner to sandbox-teleporter 

Small nit, but can this be resized to 80x80px?

> Added a "Log out" button to the main website header navigation if users are logged in

This is working for me on desktop and mobile navs

mikermcneil commented:
Once @mike-j-thomas has a working secret, he will try out the flow again, especially re: 

> (d) The white screen and broken back button behavior for the signup flow is not a good experience, but it is an unfortunate side effect of the change that was made to the plan for authentication.

eashaw commented:
@mikermcneil The Zapier webhook request has been added, and the Zap is set up and turned on.
One thing to note is that the Zapier webhook will always return a 200 response, the check for the `x-webhook-secret` header happens after Zapier receives the request (as a separate step in the zap).

mikermcneil commented on website/api/controllers/entrance/signup.js:
```diff
@@ -78,13 +83,86 @@ the account verification message.)`,
+      )
+      .timeout(5000)
+      .tolerate(['non200Response', 'requestFailed'], (err)=>{
+        sails.log.warn(`When a new user signed up for Fleet Sandbox, a lead/contact could not be verified in the CRM for this email address: ${newEmailAddress}. Raw error: ${err}`);
```
> One thing to note is that the Zapier webhook will always return a 200 response, the check for the x-webhook-secret header happens after Zapier receives the request (as a separate step in the zap).

👍 
<img width="784" alt="image" src="https://user-images.githubusercontent.com/618009/182756658-dd9dec3a-6d02-4f8f-94f4-ae1475b98f24.png">

Makes sense.  Probably worth leaving a link to [your related comment](https://github.com/fleetdm/fleet/pull/6380#issuecomment-1204395762) as a code comment here for when we're puzzling over this in the future.

eashaw commented on website/config/routes.js:
```diff
@@ -249,6 +276,7 @@ module.exports.routes = {
   'GET /docs/using-fleet/updating-fleet': '/docs/deploying/upgrading-fleet',
   'GET /blog':                   '/articles',
   'GET /brand':                  '/logos',
+  'GET /get-started':            '/try-fleet/register',
```
@noahtalerman These changes have been added to this PR

chris-mcgillicuddy commented on articles/osquery-a-tool-to-easily-ask-questions-about-operating-systems.md:
```diff
@@ -57,7 +57,7 @@ To deploy osquery at scale across your organization’s devices, an osquery mana
 ![A screenshot of the Fleet UI](../website/assets/images/articles/osquery-a-tool-to-easily-ask-questions-about-operating-systems-1-800x450@2x.png)
 
-Ready to give Fleet a try? Head over to our [get started page](https://fleetdm.com/get-started), and you can be up and running with a sandbox environment in less than 5 minutes. Alternatively, check out the [docs](https://fleetdm.com/docs) to learn how to deploy Fleet across your organization.
+Ready to give Fleet a try? Head over to our [get started page](https://fleetdm.com/try-fleet/register), and you can be up and running with a sandbox environment in less than 5 minutes. Alternatively, check out the [docs](https://fleetdm.com/docs) to learn how to deploy Fleet across your organization.
```
```suggestion
Ready to give Fleet a try? Head over to our [get started page](https://fleetdm.com/try-fleet/register), and you can be up and running with a sandbox environment in less than 5 minutes. Or you can check out the [docs](https://fleetdm.com/docs) to learn how to deploy Fleet across your organization.
```

chris-mcgillicuddy reviewed (commented):
Suggested different wording to sound more casual.

noahtalerman reviewed (approved):
Looks good! Reed and I tested redirect changes ✅

eashaw commented:
Database migrations are complete, merging now.
